### DISCUSSION TITLE
Policy CRD allows everything until first policy CR

### DISCUSSION THREAD

**Discussion by hash-d** [skupperproject/skupper#792] [Ideas]
Today, when applying a Policy CRD to a cluster, it takes effect immediately and prohibits all of the policy-enforceable features, until a Policy CR is created and policy is allowed to whatever it specifies.

This will cause disruption to an existing site, which may extend past the creation of the CR, given that some of the skupper items will be removed when policy does not allow them (skupper services, for example), and the admin will have to recreate them.

The suggestion is to make the Skupper Policy engine work in a similar fashion to how Kubernetes [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-policies) work.  There, policies are only enforced on a namespace once at least one policy applies to that namespace, otherwise all traffic is allowed.

The difference here on skupper is that the policies should be enforcing or permissive cluster-wide, instead of namespace-wide.  This would simplify applying policies for the first time:

- Install Policy CRD (no policies are in effect)
- Install all-allowing Policy CR (policies are now in effect, but this CR keeps everything up)
- Install actual policies that are desired on the cluster (the actual policies are installed, but the all-allowing one supersedes them and everything is still allowed)
- Remove the all-allowing policy (now, only the desired policies are in effect)

**Comment by hash-d** [comment]
Note #759 removes the 'extension past creation of the CR part', but still one would have some disruption when adding the CRD to the cluster.

**Reply by fgiorgetti** [reply]
It seems like a good approach, but I believe the following item is not needed:
_Install all-allowing Policy CR (policies are now in effect, but this CR keeps everything up)_

If there aren't any policies that apply to a given namespace, the policy mechanism can consider everything as allowed.
In case someone wants to enforce that unauthorized namespaces cannot have an operational skupper site, they can
simply add an empty policy that applies to all namespaces, right after installing the CRD.

**Reply by fgiorgetti** [reply]
+1 to include it in the pipeline

**Reply by hash-d** [reply]
I think what you suggested is really better.  The first time policy install would then become:

-    Install Policy CRD (no policies are in effect)
-    Install actual policies that are desired on the cluster (these CRs enforce the policies on the namespaces they specify)
-    Install an empty CR that applies to all namespaces (from this point on, policies apply to all namespaces, and the previously applied policies grant the required permissions to the namespaces they specify)

The advantage is that the actual policies can be tested before policies are enforced cluster-wise

**Reply by grs** [reply]
That third step would be important to document as best practice, as until that is in place any new namespace will be able to do anything skupper-wise.

**Reply by fgiorgetti** [reply]
One more thing to add here. If a given site has been installed by a namespace admin (no permission to read policies),
policies must be enforced, regardless of whether there are policies for the respective namespace or not.

**Reply by grs** [reply]
Why?

**Reply by fgiorgetti** [reply]
Because there is no way block the site if controller can't read policies. In this case, lack of permission to read would imply it's enabled.

**Reply by grs** [reply]
Of course, makes sense. That does make the overall picture somewhat less clear for this change in my view. (I wonder if we could add a subcommand to skupper cli that would indicate whether a given site - or perhaps even scan all sites - would be affected by policy installation?

**Reply by fgiorgetti** [reply]
That is a good point. We had a discussion about it last week and we have been requested to write a "dry-run" command
for policies, which, I believe, would do exactly what you suggested.

Along with that, I believe `skupper status` should report that a given site is _not operational_ as policies are enabled on
the cluster, but controller cannot read them. Same for `skupper init` performed by namespace admins after policy is enabled.
